astropy/__init__.py

```diff
@@ -6,6 +6,18 @@
 managing them.
 """
 
+# Prior to Astropy 3.2, astropy was imported during setup.py commands. If we are
+# in setup mode, then astropy-helpers defines an _ASTROPY_SETUP_ variable, which
+# we used to use to conditionally import C extensions for example. However, the
+# behavior of importing the package during the setup process is not good
+# practice and we therefore now explicitly prevent the package from being
+# imported in that case to prevent any regressions.
+try:
+    _ASTROPY_SETUP_
+except NameError:
+    pass
+else:
+    raise Exception("The astropy package cannot be imported during setup")
```

Comment:
Doesn't this mean that if a *different* package (e.g., anything using the affiliated package template) assigns `_ASTROPY_SETUP_` and then tries to import astropy it will fail?  I think that's not desirable, right?  Because conceivably affiliated packages (which may still use `_ASTROPY_SETUP_`) may want to import astropy at install time? If I'm understanding that right we might want to either convert this to a deprecation warning that's more like "this package should stop using `_ASTROPY_SETUP_`, see issue #xxx" (where xxx might be this, or a more specific issue in the package template?)

Also a thought for a slightly more specific exception *if* the above is wrong/rejected:
```suggestion
    raise RuntimeError("The astropy package cannot be imported during setup")
```